web3/contract.py

```diff
@@ -1442,6 +1449,82 @@ def factory(cls, class_name, **kwargs):
         return PropertyCheckingFactory(class_name, (cls,), kwargs)
 
 
+class ContractCaller:
+    """
+    An alternative Contract API.
+    This call:
+    > contract.caller({'from': eth.accounts[1], 'gas': 100000, ...}).add(2, 3)
+    is equivalent to this call in the classic contract:
+    > contract.functions.add(2, 3).call({'from': eth.accounts[1], 'gas': 100000, ...})
+    Other options for invoking this class include:
+    > contract.caller.add(2, 3)
+    or
+    > contract.caller().add(2, 3)
+    or
+    > contract.caller(transaction_dict={'from': eth.accounts[1], 'gas': 100000, ...}).add(2, 3)
+    """
+    def __init__(self,
+                 abi,
+                 web3,
+                 address,
+                 transaction_dict=None,
+                 block_identifier='latest'):
+        self.web3 = web3
+        self.address = address
+
+        if abi:
+            self.abi = abi
+
+            if transaction_dict is None:
+                transaction_dict = {}
+
+            self._functions = filter_by_type('function', self.abi)
+            for func in self._functions:
+                fn = ContractFunction.factory(
+                    func['name'],
+                    web3=self.web3,
+                    contract_abi=self.abi,
+                    address=self.address,
+                    function_identifier=func['name'])
+
+                block_id = parse_block_identifier(self.web3, block_identifier)
+                caller_method = partial(self.call_function,
+                                        fn,
+                                        transaction_dict=transaction_dict,
+                                        block_identifier=block_id)
+
+                setattr(self, func['name'], caller_method)
+
+    def __getattr__(self, function_name):
+        if '_functions' not in self.__dict__:
+            raise NoABIFunctionsFound(
+                "The abi for this contract contains no function definitions. ",
+                "Are you sure you provided the correct contract abi?"
+            )
+        elif function_name not in self.__dict__['_functions']:
+            raise MismatchedABI(
+                "The function '{}' was not found in this contract's abi. ".format(function_name),
+                "Are you sure you provided the correct contract abi?"
```

Comment:
It might be useful to include the names of all of the known functions in this message since it's very likely a user encountering this error has either mispelled a function name, or upon encountering it, will want to know what names are available.